common/lib/common-utils/src/instanceOf.ts

```diff
@@ -0,0 +1,49 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+const toString: (value: any) => string = (value) => Object.prototype.toString.call(value);
+const objectString = toString(Object.prototype);
+
+/**
+ * Plain-Old-JavaScript-Object -- as recommended by TypeScript
+ */
+export type POJO = Record<string, unknown>;
+
+/**
+ * Tests if the passed value is a plain-old-javascript-object
+ * @param value - the value to check
+ * @returns true if the value is an object, otherwise false
+ */
+export function instanceOfObject(value: any): value is POJO {
+    return value !== null && typeof value === "object";
+}
+
+/**
+ * Produces partially applied functions that test if passed values match the provided builtin instance type
+ * @param match - an instance of the builtin to match subsequent values with
+ * @returns the partially applied test function -- (value: any): value is typeof match
+ */
+export function bindInstanceOfBuiltin<T>(match: T) {
+    const compareString = toString(match);
+
+    if (compareString === objectString) {
+        throw new Error("bindInstanceOfBuiltin cannot classify '[object Object]' instances");
+    }
+
+    return (value: any): value is T => compareString === toString(value);
+}
+
+// RFC:
+// Object.freeze?  Non-configurable getters?  Individual exports?
+// How "hostile" should the environment be treated?
+export const instanceOf = {
+    Object: instanceOfObject,
+
+    ArrayBuffer: bindInstanceOfBuiltin(ArrayBuffer.prototype),
+    Error: bindInstanceOfBuiltin(new Error("N/A")),
+    Map: bindInstanceOfBuiltin(Map.prototype),
+    Set: bindInstanceOfBuiltin(Set.prototype),
+    Uint8Array: bindInstanceOfBuiltin(new Uint8Array()),
+};
```

Comment:
Maybe add `as const` here? Enforce all properties as readonly and give strictest typing possible.

Reply:
Thanks @arinwt , decided to draft an `Object.freeze` variant here that importers can opt into.  One can pretty trivially [amortize the cost of frozen object property reads](https://github.com/microsoft/FluidFramework/pull/6674/files#diff-749b6cb4032a5f88a1a59c2dcf1fe9946e053f00666d9ed7382663be201628b5R10) if so inclined.

Also on the topic of "environment hostility", @vladsud , you might be interested in the [efforts taken here](https://github.com/microsoft/FluidFramework/pull/6674/files#diff-9d8334829a7f3f5e58ab34d00d5605ac50eaf4cc9bc088c6d2677374b6950796R10-R12) (with [test automation](https://github.com/microsoft/FluidFramework/pull/6674/files#diff-b0ac7d267989473a58831f60552529e3474d43b7106a57b21821a8558055ebb1R131-R132)) to insulate this implementation from modifications to the prototype chain (subsequent to top-level script execution/import), as cautioned by http://web.mit.edu/jwalden/www/isArray.html

> relies on Object.prototype.toString and Function.prototype.call not being changed (probably a good assumption but still fragile)

Definitely interested in hearing feedback if the techniques proposed here seem too "eccentric".